validator_client/src/beacon_node_fallback.rs

```diff
@@ -475,225 +549,120 @@ impl<T: SlotClock, E: EthSpec> BeaconNodeFallback<T, E> {
-                            log,
-                            "Request to beacon node failed";
-                            "node" => $candidate.beacon_node.to_string(),
-                            "error" => ?e,
-                        );
-                        // If we have an error on this function, make the client as not-ready.
-                        //
-                        // There exists a race condition where the candidate may have been marked
-                        // as ready between the `func` call and now. We deem this an acceptable
-                        // inefficiency.
-                        if matches!(offline_on_failure, OfflineOnFailure::Yes) {
-                            $candidate.set_offline().await;
-                        }
-                        errors.push(($candidate.beacon_node.to_string(), Error::RequestFailed(e)));
-                        inc_counter_vec(&ENDPOINT_ERRORS, &[$candidate.beacon_node.as_ref()]);
-                    }
-                }
-            }};
+        for candidate in candidates.iter() {
+            futures.push(Self::run_on_candidate(
+                candidate.beacon_node.clone(),
+                &func,
+                &self.log,
+            ));
         }
+        drop(candidates);
 
-        // First pass: try `func` on all synced and ready candidates.
-        //
-        // This ensures that we always choose a synced node if it is available.
-        for candidate in &self.candidates {
-            match candidate.status(RequireSynced::Yes).await {
-                Err(e @ CandidateError::NotSynced) if require_synced == false => {
-                    // This client is unsynced we will try it after trying all synced clients
-                    retry_unsynced.push(candidate);
-                    errors.push((candidate.beacon_node.to_string(), Error::Unavailable(e)));
-                }
-                Err(e) => {
-                    // This client was not ready on the first pass, we might try it again later.
-                    to_retry.push(candidate);
-                    errors.push((candidate.beacon_node.to_string(), Error::Unavailable(e)));
-                }
-                _ => try_func!(candidate),
+        for future in futures {
+            match future.await {
+                Ok(val) => return Ok(val),
+                Err(e) => errors.push(e),
             }
         }
 
-        // Second pass: try `func` on ready unsynced candidates. This only runs if we permit
-        // unsynced candidates.
-        //
-        // Due to async race-conditions, it is possible that we will send a request to a candidate
-        // that has been set to an offline/unready status. This is acceptable.
-        if require_synced == false {
-            for candidate in retry_unsynced {
-                try_func!(candidate);
-            }
-        }
+        // Second pass. No candidates returned successfully. Try again with the same order.
```

Comment:
Minor note: this won't actually guarantee the order now we're re-acquiring the lock.
We could clone the candidate list at the beginning so we don't need the extra for loops and re-acquiring read lock to simplify this function.

Reply:
Happy for this to be addressed separately

Reply:
Raised a separate issue https://github.com/sigp/lighthouse/issues/6459

Reply:
I actually think this is a feature not a bug, this code path only runs in the event that no candidates returned successfully, so if the order gets updated for the second pass I feel like that's actually a good thing.

Although I can see the argument that no needing to acquire the read lock might be more efficient than the clone

Reply:
Yep I wasn't sure if this was a bug either, thought it could be intentional!